File: chrome-extension/src/kaggie-graph-agent/tools/rag_tool.ts
// Exact mirror of backend/src/tools/rag_tool.py
import { tool } from '@langchain/core/tools';
import { z } from 'zod';
import { globalConfig } from '../../config/globalConfig';
import { embeddingService } from '../../services/embeddingService';
export const ragTool = tool(
  async ({ query, competition_id }: { query: string; competition_id?: string }) => {
    try {
      // Make API call to backend RAG endpoint
      await globalConfig.initialize();
      const backendUrl = globalConfig.getBackendUrl();
      
      console.log('🔍 RAG Tool: Searching for:', query, 'Competition:', competition_id);
      console.log('🔍 RAG Tool: Backend URL:', backendUrl);
      const embedded_query = await embeddingService.embedText(query);
      console.log("Embedded Query", embedded_query)
      const response = await fetch(`${backendUrl}/api/rag-search`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ 
          embedding: embedded_query, 
          competition_id: competition_id || null,
          k: 4 // Default number of results
        }),
      });

      if (!response.ok) {
        const errorText = await response.text();
        console.error('🔍 RAG Tool: Backend error:', response.status, errorText);
        throw new Error(`RAG search failed: ${response.statusText} - ${errorText}`);
      }

      const response_data = await response.json();
      console.log('🔍 RAG Tool: Retrieved response:', response_data);
      console.log('🔍 RAG Tool: Response results count:', response_data.results?.length || 0);
      
      // Extract the results array from the response
      const results = response_data.results || [];
      
      // Handle the backend API response format
      if (!results || results.length === 0) {
        console.log('🔍 RAG Tool: No documents found');
        return "No relevant documents found.";
      }

      // Format results to match expected structure
      // Backend returns: { content, metadata }
      const formattedResults = results.map((doc: any, i: number) => {
        const content = doc.content || '';
        const url = doc.metadata?.url || 'No URL';
        
        return `Document ${i + 1}: ${content}, Source URL: ${url}`;
      });

      const finalResult = formattedResults.join('\n\n');
      console.log('🔍 RAG Tool: Formatted result length:', finalResult.length);
      
      return finalResult;
    } catch (error) {
      console.error('🔍 RAG Tool error:', error);
      return `RAG search error: ${error}`;
    }
  },
  {
    name: 'rag_tool',
    description: 'Search competition discussions using semantic similarity',
    schema: z.object({
      query: z.string().describe('Search query for finding relevant content'),
      competition_id: z.string().optional().describe('Target competition to filter results'),
    }),
  }
);

File: backend/tools/rag_tool.py
from langchain_core.tools import tool
from services.vector_store import vector_store_service

@tool
def rag_tool(query: str, competition_id: str = None):
    """Search competition discussions using semantic similarity.
    
    Args:
        query: Search query for finding relevant content
        competition_id: Target competition to filter results (optional)
    
    Returns:
        Formatted results with document content and source URLs
    """
    if competition_id:
        query_results = vector_store_service.search_discussions(query=query, competition_id=competition_id)
    else:
        query_results = vector_store_service.search_by_text(query=query)
    
    # Format results
    if not query_results:
        return "No relevant documents found."
    
    results = []
    for i, doc in enumerate(query_results, 1):
        url = doc.metadata.get("url", "No URL available")
        doc_type = doc.metadata.get("type", "unknown")
        title = doc.metadata.get("title", "No title")
        
        results.append(f"Document {i} [{doc_type.upper()}]: {title}\nContent: {doc.page_content[:300]}...\nSource URL: {url}")
    
    return "\n\n".join(results)

class RAGTool:
    """RAG Tool class for compatibility and easier testing"""
    
    def __init__(self):
        self.vector_store = vector_store_service
    
    def search(self, query: str, competition_id: str = None, k: int = 4):
        """Search for relevant documents using the vector store"""
        if competition_id:
            return self.vector_store.search_discussions(query=query, competition_id=competition_id, k=k)
        else:
            return self.vector_store.search_by_text(query=query, k=k)
    
    def search_competitions(self, query: str, k: int = 4):
        """Search for relevant competitions"""
        return self.vector_store.search_competitions(query=query, k=k)
    
    def search_discussions(self, query: str, competition_id: str = None, k: int = 4):
        """Search for relevant discussions"""
        return self.vector_store.search_discussions(query=query, competition_id=competition_id, k=k)